platforms/jvm/toolchains-jvm/src/main/java/org/gradle/jvm/toolchain/internal/install/DefaultJavaToolchainProvisioningService.java

```diff
@@ -111,17 +117,66 @@ public File tryInstall(JavaToolchainSpec spec) {
 
-        for (RealizedJavaToolchainRepository request : repositories) {
-            Optional<JavaToolchainDownload> download = request.getResolver().resolve(new DefaultJavaToolchainRequest(spec, buildPlatform));
-            if (download.isPresent()) {
-                Collection<Authentication> authentications = request.getAuthentications(download.get().getUri());
-                return provisionInstallation(spec, download.get().getUri(), authentications);
+        Map<String, Exception> resolveFailures = new TreeMap<>();
+        Map<String, Exception> provisioningFailures = new TreeMap<>();
+        File successfulProvisioning = null;
+        for (RealizedJavaToolchainRepository repository : repositories) {
+            JavaToolchainResolver resolver = repository.getResolver();
+            Optional<JavaToolchainDownload> download;
+            try {
+                download = resolver.resolve(new DefaultJavaToolchainRequest(spec, buildPlatform));
+            } catch (Exception e) {
+                resolveFailures.put(repository.getRepositoryName(), e);
+                continue;
+            }
+            try {
+                if (download.isPresent()) {
+                    Collection<Authentication> authentications = repository.getAuthentications(download.get().getUri());
+                    successfulProvisioning = provisionInstallation(spec, download.get().getUri(), authentications);
+                    break;
+                }
+            } catch (Exception e) {
+                provisioningFailures.put(repository.getRepositoryName(), e);
+                // continue
             }
         }
 
-        throw new ToolchainDownloadFailedException("No locally installed toolchains match and the configured toolchain download repositories aren't able to provide a match either.",
-            "Learn more about toolchain auto-detection at " + Documentation.userManual("toolchains", "sec:auto_detection").getUrl() + ".",
-            "Learn more about toolchain repositories at " + Documentation.userManual("toolchains", "sub:download_repositories").getUrl() + ".");
+        if (successfulProvisioning == null) {
+            String message = "No locally installed toolchains match and the configured toolchain download repositories aren't able to provide a match either." +
+                (hasFailures(resolveFailures, provisioningFailures) ? " " + formatFailures(resolveFailures, provisioningFailures) : "");
+            throw new ToolchainDownloadFailedException(message,
+                "Learn more about toolchain auto-detection at " + Documentation.userManual("toolchains", "sec:auto_detection").getUrl() + ".",
+                "Learn more about toolchain repositories at " + Documentation.userManual("toolchains", "sub:download_repositories").getUrl() + ".");
+        } else {
+            if (hasFailures(resolveFailures, provisioningFailures)) {
+                LOGGER.warn(formatFailures(resolveFailures, provisioningFailures));
```

Comment:
🤔 Logging just a message an WARN level sounds good but it would be good to provide a way for users to look at the stack traces of the underlying exceptions. Maybe build an exception with the underlying ones as suppressed, log this to a lower logging level and add a hint to the warning saying that.